[Bug]: Pyparsing 3.1 breaks tests

### Bug summary

Pyparsing 3.1.0 was released today and something has changed which causes quite a bit of test failures.

List of changes https://github.com/pyparsing/pyparsing/releases/tag/3.1.0 (not obvious to me what caused it).

### Code for reproduction

```python
Run the tests
```


### Actual outcome

```
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[hspace without value]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[hspace with invalid value]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[function without space]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[accent without space]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[frac without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[frac with empty parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[binom without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[binom with empty parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[genfrac without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[genfrac with empty parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[sqrt without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[sqrt with invalid value]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[overline without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[overline with empty parameter]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[left with invalid delimiter]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[right with invalid delimiter]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[unclosed parentheses with sizing]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[unclosed parentheses without sizing]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[dfrac without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[dfrac with empty parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[overset without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[underset without parameters]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[unknown symbol]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[double superscript]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[double subscript]
FAILED lib/matplotlib/tests/test_mathtext.py::test_mathtext_exceptions[super on sub without braces]
FAILED lib/matplotlib/tests/test_text.py::test_parse_math - AssertionError: R...
FAILED lib/matplotlib/tests/test_text.py::test_parse_math_rcparams - Assertio...
```

### Expected outcome

No failures

### Additional information

I guess a two-stage fix:
1) Pin the version to `<3.1`
2) Fix the actual problem and unpin

### Operating system

_No response_

### Matplotlib Version

main-branch

### Matplotlib Backend

_No response_

### Python version

_No response_

### Jupyter version

_No response_

### Installation

None

These are the failing tests:
https://github.com/matplotlib/matplotlib/blob/060992a18e1f9d6a370f73c02583f314631bc96f/lib/matplotlib/tests/test_mathtext.py#L279-L346

Here is the full traceback of one of them:
```
――――――――――――――――――――――――――――――――――――― test_mathtext_exceptions[double subscript] ――――――――――――――――――――――――――――――――――――――

math = '$a_2_2$', msg = 'Double subscript'

    @pytest.mark.parametrize(
        'math, msg',
        [
            (r'$\hspace{}$', r'Expected \hspace{space}'),
            (r'$\hspace{foo}$', r'Expected \hspace{space}'),
            (r'$\sinx$', r'Unknown symbol: \sinx'),
            (r'$\dotx$', r'Unknown symbol: \dotx'),
            (r'$\frac$', r'Expected \frac{num}{den}'),
            (r'$\frac{}{}$', r'Expected \frac{num}{den}'),
            (r'$\binom$', r'Expected \binom{num}{den}'),
            (r'$\binom{}{}$', r'Expected \binom{num}{den}'),
            (r'$\genfrac$',
             r'Expected \genfrac{ldelim}{rdelim}{rulesize}{style}{num}{den}'),
            (r'$\genfrac{}{}{}{}{}{}$',
             r'Expected \genfrac{ldelim}{rdelim}{rulesize}{style}{num}{den}'),
            (r'$\sqrt$', r'Expected \sqrt{value}'),
            (r'$\sqrt f$', r'Expected \sqrt{value}'),
            (r'$\overline$', r'Expected \overline{body}'),
            (r'$\overline{}$', r'Expected \overline{body}'),
            (r'$\leftF$', r'Expected a delimiter'),
            (r'$\rightF$', r'Unknown symbol: \rightF'),
            (r'$\left(\right$', r'Expected a delimiter'),
            # PyParsing 2 uses double quotes, PyParsing 3 uses single quotes and an
            # extra backslash.
            (r'$\left($', re.compile(r'Expected ("|\'\\)\\right["\']')),
            (r'$\dfrac$', r'Expected \dfrac{num}{den}'),
            (r'$\dfrac{}{}$', r'Expected \dfrac{num}{den}'),
            (r'$\overset$', r'Expected \overset{annotation}{body}'),
            (r'$\underset$', r'Expected \underset{annotation}{body}'),
            (r'$\foo$', r'Unknown symbol: \foo'),
            (r'$a^2^2$', r'Double superscript'),
            (r'$a_2_2$', r'Double subscript'),
            (r'$a^2_a^2$', r'Double superscript'),
        ],
        ids=[
            'hspace without value',
            'hspace with invalid value',
            'function without space',
            'accent without space',
            'frac without parameters',
            'frac with empty parameters',
            'binom without parameters',
            'binom with empty parameters',
            'genfrac without parameters',
            'genfrac with empty parameters',
            'sqrt without parameters',
            'sqrt with invalid value',
            'overline without parameters',
            'overline with empty parameter',
            'left with invalid delimiter',
            'right with invalid delimiter',
            'unclosed parentheses with sizing',
            'unclosed parentheses without sizing',
            'dfrac without parameters',
            'dfrac with empty parameters',
            'overset without parameters',
            'underset without parameters',
            'unknown symbol',
            'double superscript',
            'double subscript',
            'super on sub without braces'
        ]
    )
    def test_mathtext_exceptions(math, msg):
        parser = mathtext.MathTextParser('agg')
        match = re.escape(msg) if isinstance(msg, str) else msg
        with pytest.raises(ValueError, match=match):
>           parser.parse(math)

test_mathtext.py:346:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <matplotlib.mathtext.MathTextParser object at 0x0000018B5CC266D0>, s = '$a_2_2$', dpi = 72, prop = None

    def parse(self, s, dpi=72, prop=None):
        """
        Parse the given math expression *s* at the given *dpi*.  If *prop* is
        provided, it is a `.FontProperties` object specifying the "default"
        font to use in the math expression, used for all non-math text.

        The results are cached, so multiple calls to `parse`
        with the same expression should be fast.

        Depending on the *output* type, this returns either a `VectorParse` or
        a `RasterParse`.
        """
        # lru_cache can't decorate parse() directly because prop
        # is mutable; key the cache using an internal copy (see
        # text._get_text_metrics_with_cache for a similar case).
        prop = prop.copy() if prop is not None else None
>       return self._parse_cached(s, dpi, prop)

..\mathtext.py:77:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <matplotlib.mathtext.MathTextParser object at 0x0000018B5CC266D0>, s = '$a_2_2$', dpi = 72
prop = <matplotlib.font_manager.FontProperties object at 0x0000018B5CA92C70>

    @functools.lru_cache(50)
    def _parse_cached(self, s, dpi, prop):
        from matplotlib.backends import backend_agg

        if prop is None:
            prop = FontProperties()
        fontset_class = _api.check_getitem(
            self._font_type_mapping, fontset=prop.get_math_fontfamily())
        load_glyph_flags = {
            "vector": LOAD_NO_HINTING,
            "raster": backend_agg.get_hinting_flag(),
        }[self._output_type]
        fontset = fontset_class(prop, load_glyph_flags)

        fontsize = prop.get_size_in_points()

        if self._parser is None:  # Cache the parser globally.
            self.__class__._parser = _mathtext.Parser()

>       box = self._parser.parse(s, fontset, fontsize, dpi)

..\mathtext.py:98:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <matplotlib._mathtext.Parser object at 0x0000018B57DB44C0>, s = '$a_2_2$'
fonts_object = <matplotlib._mathtext.BakomaFonts object at 0x0000018B5CA92910>, fontsize = 12.0, dpi = 72

    def parse(self, s, fonts_object, fontsize, dpi):
        """
        Parse expression *s* using the given *fonts_object* for
        output, at the given *fontsize* and *dpi*.

        Returns the parse tree of `Node` instances.
        """
        self._state_stack = [
            ParserState(fonts_object, 'default', 'rm', fontsize, dpi)]
        self._em_width_cache = {}
        try:
            result = self._expression.parseString(s)
        except ParseBaseException as err:
            # explain becomes a plain method on pyparsing 3 (err.explain(0)).
>           raise ValueError("\n" + ParseException.explain(err, 0)) from None
E           ValueError:
E
E           ^
E           ParseFatalException: Expected token  (at char 0), (line:1, col:1)

..\_mathtext.py:2056: ValueError

During handling of the above exception, another exception occurred:

math = '$a_2_2$', msg = 'Double subscript'

    @pytest.mark.parametrize(
        'math, msg',
        [
            (r'$\hspace{}$', r'Expected \hspace{space}'),
            (r'$\hspace{foo}$', r'Expected \hspace{space}'),
            (r'$\sinx$', r'Unknown symbol: \sinx'),
            (r'$\dotx$', r'Unknown symbol: \dotx'),
            (r'$\frac$', r'Expected \frac{num}{den}'),
            (r'$\frac{}{}$', r'Expected \frac{num}{den}'),
            (r'$\binom$', r'Expected \binom{num}{den}'),
            (r'$\binom{}{}$', r'Expected \binom{num}{den}'),
            (r'$\genfrac$',
             r'Expected \genfrac{ldelim}{rdelim}{rulesize}{style}{num}{den}'),
            (r'$\genfrac{}{}{}{}{}{}$',
             r'Expected \genfrac{ldelim}{rdelim}{rulesize}{style}{num}{den}'),
            (r'$\sqrt$', r'Expected \sqrt{value}'),
            (r'$\sqrt f$', r'Expected \sqrt{value}'),
            (r'$\overline$', r'Expected \overline{body}'),
            (r'$\overline{}$', r'Expected \overline{body}'),
            (r'$\leftF$', r'Expected a delimiter'),
            (r'$\rightF$', r'Unknown symbol: \rightF'),
            (r'$\left(\right$', r'Expected a delimiter'),
            # PyParsing 2 uses double quotes, PyParsing 3 uses single quotes and an
            # extra backslash.
            (r'$\left($', re.compile(r'Expected ("|\'\\)\\right["\']')),
            (r'$\dfrac$', r'Expected \dfrac{num}{den}'),
            (r'$\dfrac{}{}$', r'Expected \dfrac{num}{den}'),
            (r'$\overset$', r'Expected \overset{annotation}{body}'),
            (r'$\underset$', r'Expected \underset{annotation}{body}'),
            (r'$\foo$', r'Unknown symbol: \foo'),
            (r'$a^2^2$', r'Double superscript'),
            (r'$a_2_2$', r'Double subscript'),
            (r'$a^2_a^2$', r'Double superscript'),
        ],
        ids=[
            'hspace without value',
            'hspace with invalid value',
            'function without space',
            'accent without space',
            'frac without parameters',
            'frac with empty parameters',
            'binom without parameters',
            'binom with empty parameters',
            'genfrac without parameters',
            'genfrac with empty parameters',
            'sqrt without parameters',
            'sqrt with invalid value',
            'overline without parameters',
            'overline with empty parameter',
            'left with invalid delimiter',
            'right with invalid delimiter',
            'unclosed parentheses with sizing',
            'unclosed parentheses without sizing',
            'dfrac without parameters',
            'dfrac with empty parameters',
            'overset without parameters',
            'underset without parameters',
            'unknown symbol',
            'double superscript',
            'double subscript',
            'super on sub without braces'
        ]
    )
    def test_mathtext_exceptions(math, msg):
        parser = mathtext.MathTextParser('agg')
        match = re.escape(msg) if isinstance(msg, str) else msg
        with pytest.raises(ValueError, match=match):
>           parser.parse(math)
E           AssertionError: Regex pattern did not match.
E            Regex: 'Double\\ subscript'
E            Input: '\n\n^\nParseFatalException: Expected token  (at char 0), (line:1, col:1)'

test_mathtext.py:346: AssertionError

```

All other looks very similar.

--- a/environment.yml
+++ b/environment.yml
@@ -19,7 +19,7 @@ dependencies:
   - pillow>=6.2
   - pybind11>=2.6.0
   - pygobject
-  - pyparsing!=3.1.0
+  - pyparsing>=2.3.1
   - pyqt
   - python-dateutil>=2.1
   - setuptools

--- a/lib/matplotlib/tests/test_mathtext.py
+++ b/lib/matplotlib/tests/test_mathtext.py
@@ -9,12 +9,17 @@ from xml.etree import ElementTree as ET
 from typing import Any
 
 import numpy as np
+from packaging.version import parse as parse_version
+import pyparsing
 import pytest
+
 
 import matplotlib as mpl
 from matplotlib.testing.decorators import check_figures_equal, image_comparison
 import matplotlib.pyplot as plt
 from matplotlib import mathtext, _mathtext
+
+pyparsing_version = parse_version(pyparsing.__version__)
 
 
 # If test is removed, use None as placeholder
@@ -279,6 +284,9 @@ def test_fontinfo():
     assert table['version'] == (1, 0)
 
 
+# See gh-26152 for more context on this xfail
+@pytest.mark.xfail(pyparsing_version.release == (3, 1, 0),
+                   reason="Error messages are incorrect for this version")
 @pytest.mark.parametrize(
     'math, msg',
     [

--- a/lib/matplotlib/tests/test_text.py
+++ b/lib/matplotlib/tests/test_text.py
@@ -4,6 +4,8 @@ import warnings
 
 import numpy as np
 from numpy.testing import assert_almost_equal
+from packaging.version import parse as parse_version
+import pyparsing
 import pytest
 
 import matplotlib as mpl
@@ -15,6 +17,8 @@ import matplotlib.transforms as mtransforms
 from matplotlib.testing.decorators import check_figures_equal, image_comparison
 from matplotlib.testing._markers import needs_usetex
 from matplotlib.text import Text, Annotation
+
+pyparsing_version = parse_version(pyparsing.__version__)
 
 
 @image_comparison(['font_styles'])
@@ -818,6 +822,9 @@ def test_unsupported_script(recwarn):
          (r"Matplotlib currently does not support Bengali natively.",)])
 
 
+# See gh-26152 for more information on this xfail
+@pytest.mark.xfail(pyparsing_version.release == (3, 1, 0),
+                   reason="Error messages are incorrect with pyparsing 3.1.0")
 def test_parse_math():
     fig, ax = plt.subplots()
     ax.text(0, 0, r"$ \wrong{math} $", parse_math=False)
@@ -828,6 +835,9 @@ def test_parse_math():
         fig.canvas.draw()
 
 
+# See gh-26152 for more information on this xfail
+@pytest.mark.xfail(pyparsing_version.release == (3, 1, 0),
+                   reason="Error messages are incorrect with pyparsing 3.1.0")
 def test_parse_math_rcparams():
     # Default is True
     fig, ax = plt.subplots()

--- a/setup.py
+++ b/setup.py
@@ -335,7 +335,7 @@ setup(  # Finally, pass this all along to setuptools to do the heavy lifting.
         "numpy>=1.21",
         "packaging>=20.0",
         "pillow>=6.2.0",
-        "pyparsing>=2.3.1,<3.1",
+        "pyparsing>=2.3.1",
         "python-dateutil>=2.7",
     ] + (
         # Installing from a git checkout that is not producing a wheel.

